### PR TITLE
adding intralist diversity metric

### PR DESCRIPTION
**Pull request by wddcheng**

### CODE REVIEW

File: tests/test_fairness_binary.py
Comment (dorukkilitcioglu) [inline tests/test_fairness_binary.py:853]: Wow, thanks for identifying and fixing this! I wonder what was being returned from this function originally

File: jurity/recommenders/intralist_diversity.py
Comment (bkleyn) [inline jurity/recommenders/intralist_diversity.py:1]: Would it make sense to combine `intralist_diversity.py` and `interlist_diversity.py` into a single `diversity.py` module?

File: jurity/recommenders/intralist_diversity.py
Comment (bkleyn) [inline jurity/recommenders/intralist_diversity.py:5]: Tuple not used

File: jurity/recommenders/intralist_diversity.py
Comment (bkleyn) [inline jurity/recommenders/intralist_diversity.py:10]: tocsr not used

File: jurity/recommenders/intralist_diversity.py
Comment (bkleyn) [inline jurity/recommenders/intralist_diversity.py:113]: Extra #

File: tests/test_reco_diversity.py
Comment (bkleyn) [inline tests/test_reco_diversity.py:5]: Can drop `RankingMetrics` in line 10

**Review by bkleyn — COMMENTED** [review]
Very well done!

Some minor comments:
- I think there could be an opportunity to combine the inter- and intra- diversity modules into a single `diversity.py` module
- Few minor formatting things that can be fixed
- Remember to update version to 1.3.2 in jurity/_version.py

File: tests/test_fairness_binary.py
Comment (wddcheng) [reply tests/test_fairness_binary.py:853]: thank you! looks like a ndarray that was returned

File: jurity/recommenders/intralist_diversity.py
Comment (skadio) [reply jurity/recommenders/intralist_diversity.py:1]: that makes sense to me.

File: tests/test_fairness_binary.py
Comment (skadio) [reply tests/test_fairness_binary.py:853]: nice catch!

File: jurity/recommenders/intralist_diversity.py
Comment (skadio) [reply jurity/recommenders/intralist_diversity.py:1]: And then there will be a one-to-one mapping between diversity.py and test_reco_diversity.py

Comment (takojunior) [reply jurity/recommenders/intralist_diversity.py:1]: I think there is the opportunity to combine the two into a single module, for example, in ctr, there are `ips` and `dr` metrics. One thing that might be dangling for the two is they have some specific args, like intralist will need item_features. So would like to hear more comments.

**Review by skadio — APPROVED** [review]
Thank you @wddcheng  for this new functionality! It strengthens our library and will drastically improve recwiser pipeline. 

Minor comments: 
- please update Readme.MD accordingly and the usage examples there
- check with them on what Docs need updating, and how to regenerate the documentation upon changes. 

Thank you ! 
Serdar

File: jurity/recommenders/intralist_diversity.py
Comment (takojunior) [inline jurity/recommenders/intralist_diversity.py:115]: This function can also take `n_jobs` argument. Given the number of items and feature dimension usually would not be too large to handle, this might not be needed. But if inter and intra get merged together, this is a parameter shared in common.

File: jurity/recommenders/intralist_diversity.py
Comment (takojunior) [inline jurity/recommenders/intralist_diversity.py:220]: For intra-list, would it make sense to be able to support batch_accumulate, since the results can be calculated for each user in one batch and then the user-level results can be stacked with the following batches. Thoughts? @dorukkilitcioglu @bkleyn

File: jurity/recommenders/intralist_diversity.py
Comment (takojunior) [inline jurity/recommenders/intralist_diversity.py:234]: Given `item_features` is a new argument, it might also have a check

File: tests/test_reco_diversity.py
Comment (takojunior) [inline tests/test_reco_diversity.py:192]: Nice tests! One thing might be updated is to capture the warning message if any, to ease the tests. That is also an update I would like to have for inter-list.

**Review by takojunior — COMMENTED** [review]
PR looks great to me! The implementation of intra-list computation is nice. I left some minor comments. 

We can connect on the updates to the docs and the warning messages that we might want to capture in the unit tests. Thanks! Great work!

File: jurity/recommenders/intralist_diversity.py
Comment (wddcheng) [reply jurity/recommenders/intralist_diversity.py:5]: nice catch!

File: jurity/recommenders/intralist_diversity.py
Comment (wddcheng) [reply jurity/recommenders/intralist_diversity.py:10]: nice catch!

File: tests/test_reco_diversity.py
Comment (wddcheng) [reply tests/test_reco_diversity.py:5]: nice catch!

File: jurity/recommenders/intralist_diversity.py
Comment (wddcheng) [reply jurity/recommenders/intralist_diversity.py:110]: yes, you are right! let me change that to <

File: jurity/recommenders/intralist_diversity.py
Comment (dorukkilitcioglu) [reply jurity/recommenders/intralist_diversity.py:1]: I think having them as two separate metrics (instead of CTR with 3 different methods) make more sense. `ips` and `dr` are different methods of calculating CTR, whereas interlist and intralist diversity fundamentally sounds like different metrics to me.

That being said, I think they can share the backend (ie an abstract class like `list_diversity`).

File: jurity/recommenders/intralist_diversity.py
Comment (dorukkilitcioglu) [reply jurity/recommenders/intralist_diversity.py:220]: Ah, I totally missed that this could be used. Yeah, we can support batch_accumulate because the calculations are done for the items per user, without users interacting with each other.

Comment (bkleyn) [reply jurity/recommenders/intralist_diversity.py:220]: Yes, batch accumulate should be valid for intra-list.

File: jurity/recommenders/intralist_diversity.py
Comment (wddcheng) [reply jurity/recommenders/intralist_diversity.py:1]: @dorukkilitcioglu I think that's a good idea. so based on the discussions, the question is whether to 1. merge them into one class/one metric with different methods or 2. keep them separate but use a common base class, or 3. keep separate metrics but use one script. what do we think? @skadio @dorukkilitcioglu @takojunior @bkleyn @filip-michalsky

**Review by filip-michalsky — APPROVED** [review]
Nice work, @wddcheng ! I also confirmed all unit tests passed on my local.

File: jurity/recommenders/intralist_diversity.py
Comment (skadio) [reply jurity/recommenders/intralist_diversity.py:1]: I see where Doruk is coming from and I don't disagree with that. Still, I vote for Option 1 (merging them together).

The tricky bit with libraries, such as Jurity, that tries to do too many different things (such as fairness eval, bias mitigate, reco eval, classification, uplift perf, oh my..) is that the code comes very fragmented. 

As such, whenever there is a chance to modularize it and bundle components together, it helps the overall library to be more succinct. Xin's CTR example is a good one. Likewise, are you interested in diversity metrics for recommenders? Go check jurity/reco/diversity.py and see jurity/test/test_reco_diversity.py. 

Once combined under diversity.py, the implementation can be completely independent (option 2 above) OR using a base (option 3 above). 

On that choice, I am indifferent with slightly leaning toward Option 2 for now. I am not aware of any more upcoming diversity metric implementations that can benefit from extracting such a base class. This can be done in the future, **IF**, we need one more implementation and find ourselves repeating the code again --I doubt this case will happen any time soon. So I will leave that to Xin/Du on how they want it to be implemented at the moment.

Comment (dorukkilitcioglu) [reply jurity/recommenders/intralist_diversity.py:1]: Totally agreed on that very last point - this is already a great contribution and it will be great to finally have it in Jurity regardless of the design choices on the backend (and I'm also happy to leave it to the authors of the diversity metric implementations).

@skadio Just to confirm your view, between the two options below, would you pick B?
```python
# Option A
m1 = DiversityRecoMetrics.InterListDiversity(...)
m2 = DiversityRecoMetrics.IntraListDiversity(...)

# Option B
m1 = DiversityRecoMetrics.ListDiversity(method='interlist', ...)
m2 = DiversityRecoMetrics.ListDiversity(method='intralist', ...)
```

Comment (skadio) [reply jurity/recommenders/intralist_diversity.py:1]: It is possible that I missed something here. 

I thought we all agree on Option A, the merge comment was only about pushing the code into a single diversity.py file. 

If we can’t get Option A —the discard my comments :)

Comment (filip-michalsky) [reply jurity/recommenders/intralist_diversity.py:1]: Pushing both classes to a single file makes sense!

**Comment by wddcheng** [comment]
@skadio @dorukkilitcioglu @takojunior @bkleyn @filip-michalsky Just made another commit for the update!

**Comment by skadio** [comment]
@wddcheng I pushed an update to fix style/formating issues such as dangling whitespaces etc. PyCharm underlines such PEP-8 standards and CTRL+ALT+L (or Code -> Reformat) updates it too. Please check with Doruk/Xin/I to learn more. 

Let's make sure to update the documentation under about_reco.rst and recompile the docs/html. Then, in README please add this bullet under Recommender Metrics for the new metric.

Good to go after Doc update.

**Comment by wddcheng** [comment]
needs to reopen for further editing

**Pull request merged**
